### PR TITLE
Allow `add_season_year()` to optionally send spans backwards

### LINKED ISSUE #4877: `iris.coord_categorisation.add_season_year_preceding()`
## ✨ Feature Request
Add an alternative to `iris.coord_categorisation.add_season_year()` which assigns seasons which span two years to the preceding year instead of the following year.

## Motivation
`iris.coord_categorisation.add_season_year()` is an excellent feature and very handy for comparing multiple timeseries datasets. However, it has an inherent assumption that a season which spans two years (i.e. includes December and January) will be assigned to the later rather than the earlier year. We have had a usecase where we want to compare the effects of monthly climate indices to several months of meteorological data **after** the climate index month (since we know that specific climate indices affect the weather variables we are analysing). 

## Proposed change
Proposed change/addition to `iris.coord_categorisation` (which essentially just removes the use of `_month_year_adjusts()`):
```python
def add_season_year_preceding(
    cube, coord, name="season_year", seasons=("djf", "mam", "jja", "son")
):
    """
    Add a categorical year-of-season coordinate, with user specified
    seasons. Any season spanning two years will be assigned the preceding year.
    Args:
    * cube (:class:`iris.cube.Cube`):
        The cube containing 'coord'. The new coord will be added into
        it.
    * coord (:class:`iris.coords.Coord` or string):
        Coordinate in 'cube', or its name, representing time.
    Kwargs:
    * name (string):
        Name of the created coordinate. Defaults to "season_year".
    * seasons (:class:`list` of strings):
        List of seasons defined by month abbreviations. Each month must
        appear once and only once. Defaults to standard meteorological
        seasons ('djf', 'mam', 'jja', 'son').
    """
    # Check that the seasons are valid.
    _validate_seasons(seasons)

    # Define a categorisation function.
    def _season_year(coord, value):
        dt = _pt_date(coord, value)
        year = dt.year
        return year

    # Apply the categorisation.
    add_categorised_coord(cube, name, coord, _season_year)
```
</details>

### PR DESCRIPTION
**Pull request by trexfeathers**
### Description

Closes #4877.

The ability for `add_season_year()` to assign seasons that span year boundaries (e.g. the standard `djf` Winter) to be assigned to the preceding year rather than the following year. #4877 suggested adding an alternative function, but it seemed more appropriate to just add another parameter to `add_season_year()`.

I also took the opportunity to modernise the docstring, and the relevant test module.

### To-do

- [x] Write a proper description
- [x] What's New entry

---
[Consult Iris pull request check list]( https://scitools-iris.readthedocs.io/en/latest/developers_guide/contributing_pull_request_checklist.html)

### CODE REVIEW

**Review by stephenworsley — CHANGES_REQUESTED** [review]
Good job, I appreciate the diligence moving the tests to pytest format. Just a couple minor notes here.

File: lib/iris/coord_categorisation.py [hunk 1]
```diff
@@ -383,34 +392,40 @@ def _season_number(coord, value):
 
 
 def add_season_year(
-    cube, coord, name="season_year", seasons=("djf", "mam", "jja", "son")
+    cube,
+    coord,
+    name="season_year",
+    seasons=("djf", "mam", "jja", "son"),
+    send_spans_backwards=False,
```
[the remark this answers is not in the document]
Comment (trexfeathers) [reply lib/iris/coord_categorisation.py]: Yeah this was a real corker - how to be clear without writing a sentence. Glad we came up with something acceptable together offline.

be6c304331a99c94a6774c9104492dff6bac2854

File: lib/iris/coord_categorisation.py [hunk 2]
```diff
@@ -271,20 +271,29 @@ def _validate_seasons(seasons):
...
-    year or is in the start of a season that counts towards the next
-    year.
+    These adjustments ensure that no season spans two years by assigning months
+    to the **next** year (send_spans_backwards is False) or the **previous**
+    year (send_spans_backwards is True). E.g. Winter - djf: either assign Dec
+    to the next year, or Jan and Feb to the previous year.
 
     """
-    month_year_adjusts = [None, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
+    month_year_adjusts = np.array([None, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])
```
[the remark this answers is not in the document]
Comment (trexfeathers) [reply lib/iris/coord_categorisation.py]: 84ad6618295742760ca9487f92035925bea23e93

**Review by stephenworsley — APPROVED** [review]
Looks good to me.

**Pull request merged**